extensions/amp-ad/0.1/amp-ad-custom.js

```diff
@@ -0,0 +1,145 @@
+    this.slot_ = element.getAttribute('data-slot');
+
+    /** @private {boolean} True to suppress template checks and rendering:
+     * use this during testing, since our test harness does not yet support
+     * templates.
+     */
+    this.testMode_ = element.getAttribute('data-test');
+  }
+
+  /** @override */
+  getPriority() {
+    // Loads ads after other content.
+    return 2;
+  }
+
+  /** @override **/
+  isLayoutSupported(layout) {
+    /** @TODO Add proper support for more layouts, and figure out which ones
+     *  we're permitting */
+    return isLayoutSizeDefined(layout);
+  }
+
+  buildCallback() {
+    // Ensure that there are templates in this ad
+    const templates = this.element.querySelectorAll('template');
+    if (!this.testMode_) {
+      user().assert(templates.length > 0, 'Missing template in custom ad');
+    }
+    // And ensure that the slot value is legal
+    user().assert(this.slot_ === null || this.slot_.match(/^[0-9a-z]+$/),
+        'custom ad slot should be alphanumeric: ' + this.slot_);
+  }
+
+  /**
+   * @private getFullUrl_ Get a URL which includes a parameter indicating
+   * all slots to be fetched from this web server URL
+   * @returns {String} The URL with the "ampslots" parameter appended
+   */
+  getFullUrl_() {
+    // If this ad doesn't have a slot defined, just return the base URL
+    if (this.slot_ === null) {
+      return this.url_;
+    }
+    if (ampCustomadFullUrls === null) {
+      // The array of ad urls has not yet been built, do so now.
+      ampCustomadFullUrls = {};
+      const slots = {};
+
+      // Get the parent body of this amp-ad element. It could be the body of
+      // the main document, or it could be an enclosing iframe.
+      const body = ancestorElementsByTag(this.element, 'BODY')[0];
+      const elements = body.querySelectorAll('amp-ad[type=custom]');
+      for (let index = 0; index < elements.length; index++) {
+        const elem = elements[index];
+        const url = elem.getAttribute('data-url');
+        const slotId = elem.getAttribute('data-slot');
+        if (slotId !== null) {
+          if (!(url in slots)) {
+            slots[url] = [];
+          }
+          slots[url].push(encodeURIComponent(slotId));
+        }
+      }
+      for (const baseUrl in slots) {
+        ampCustomadFullUrls[baseUrl] = addParamToUrl(baseUrl, 'ampslots',
+          slots[baseUrl].join(','));
+      }
+    }
+    return ampCustomadFullUrls[this.url_];
+  }
+
+  /** @override */
+  layoutCallback() {
+    if (this.testMode_) {
+      // Our test harness can't currently deal with templates, so don't bother
+      // attempting layout. At some point, it would be good to fix this.
+      return Promise.resolve();
+    }
+    // If this promise has no URL yet, create one for it.
+    if (!(this.url_ in ampCustomadXhrPromises)) {
+      // Here is a promise that will return the data for this URL
+      ampCustomadXhrPromises[this.url_] = xhrFor(this.win).fetchJson(
```

Comment:
one thing here: i think we should use fullUrl as the key of the map here. Otherwise, the case of "2 slots with same URL, but one without slot ID" will lead to an indeterministic state.